[rebranch] Finish off memory effects change

bnbarham opened the pull request:
`ReadOnly`/`ArgMemOnly` were mostly moved over, but a few were missed. Update them all. Also default to `unknown` for no memory effects rather than none (ie. we should be conservative).

bnbarham committed:
[rebranch] Finish off memory effects change

`ReadOnly`/`ArgMemOnly` were mostly moved over, but a few were missed.
Update them all. Also default to `unknown` for no memory effects rather
than none (ie. we should be conservative).

bnbarham commented:
@swift-ci please test

bnbarham commented:
Build should be successful now, but I expect many test failures 😅

bnbarham commented:
@swift-ci please test

etcwilde commented on lib/IRGen/IRGenModule.cpp:
```diff
@@ -1104,7 +1104,7 @@ void IRGenModule::registerRuntimeEffect(ArrayRef<RuntimeEffect> effect,
 #define EFFECT(...) { __VA_ARGS__ }
 #define NO_MEMEFFECTS                                                          \
-  { llvm::MemoryEffects::none() }
+  { llvm::MemoryEffects::unknown() }
```
I defer to @aschwaighofer on how he wants to handle this.

bnbarham replied:
Yeah, me too :). But as far as I understand it, this matches the behavior we had. `unknown` is the equivalent to not specifying where as `none` is `ReadNone` (you can see this in the test updates in https://reviews.llvm.org/D135780).

To put another way:
  * `ReadNone` -> `MemoryEffects::none()`
  * `ReadOnly` -> `MemoryEffects::readOnly()`
  * `WriteOnly` -> `MemoryEffects::writeOnly()`
  * None specified -> `MemoryEffects::unknown()`

etcwilde replied:
Yep that looks good.

```c++
    // Upgrade old memory attributes.
    MemoryEffects ME = MemoryEffects::unknown();
    if (Attrs & (1ULL << 9)) {
      // ReadNone
      Attrs &= ~(1ULL << 9);
      ME &= MemoryEffects::none();
    }
    if (Attrs & (1ULL << 10)) {
      // ReadOnly
      Attrs &= ~(1ULL << 10);
      ME &= MemoryEffects::readOnly();
    }
```

etcwilde reviewed (approved):
Other than changing to unknown memory effects from no memory effects, this looks fine to me. I'll defer to Arnold for that part though.

aschwaighofer reviewed (approved):
LGTM

I wonder whether we should replace NO_MEMEFFECTS by UNKNOWN_MEMEFFECTS in a followup for clarity?

etcwilde commented:
> I wonder whether we should replace `NO_MEMEFFECTS` by `UNKNOWN_MEMEFFECTS` in a followup for clarity?

That seems reasonable, I was just matching up with `NO_ATTRS` and the other "none" macros already in the file.

bnbarham commented:
> I wonder whether we should replace NO_MEMEFFECTS by UNKNOWN_MEMEFFECTS in a followup for clarity?

Heh, I actually did that but then reverted it. I can just do it in this one.

bnbarham committed:
[rebranch] Rename NO_MEMEFFECTS to UNKNOWN_MEMEFFECTS

`NO_MEMEFFECTS` is quite similar to "none", which isn't the same as
unknown in this context. Rename to `UNKNOWN_MEMEFFECTS` for clarity.
